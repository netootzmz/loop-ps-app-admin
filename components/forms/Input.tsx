import React, {
  DetailedHTMLProps,
  FC,
  InputHTMLAttributes,
  TextareaHTMLAttributes,
  MouseEvent,
  useEffect,
  useState,
} from "react";
import svgs from "../../helpers/svgs";
import SvgWrapper from "../SvgWrapper";
import PasswordValidation from "./PasswordValidation";
import { useRef } from "react";
import Swal from "sweetalert2";

const Input: FC<
  DetailedHTMLProps<
    | InputHTMLAttributes<HTMLInputElement>
    | TextareaHTMLAttributes<HTMLTextAreaElement>,
    HTMLInputElement | HTMLTextAreaElement
  > & {
    helpText?: string;
    error?: string;
    svgId?: string;
    containerClassName?: string;
    passwordValidation?: boolean;
    inline?: boolean;
    full?: boolean;
    textarea?: boolean;
    copyToClip?: boolean;
    search?: boolean;
    submitFn?: (e: MouseEvent<HTMLButtonElement>) => void;
    helpColor?: string;
  }
> = ({
  error,
  helpText,
  svgId,
  containerClassName,
  passwordValidation = false,
  inline = false,
  full = false,
  textarea = false,
  copyToClip = false,
  search = false,
  submitFn = () => {},
  helpColor = '#f2711c',
  ...props
}) => {
  const valid = !!!error;
  const [countClass, setCountClass] = useState("input__count");

  const [show, setShow] = useState(false);
  const [type, setType] = useState(
    (
      props as DetailedHTMLProps<
        InputHTMLAttributes<HTMLInputElement>,
        HTMLInputElement
      >
    ).type || ""
  );

  const handleShow = (
    e: MouseEvent<HTMLButtonElement, globalThis.MouseEvent>
  ) => {
    e.preventDefault();
    setShow(true);
    setType("text");
  };
  const handleUnshow = (
    e: MouseEvent<HTMLButtonElement, globalThis.MouseEvent>
  ) => {
    e.preventDefault();
    setShow(false);
    setType(
      (
        props as DetailedHTMLProps<
          InputHTMLAttributes<HTMLInputElement>,
          HTMLInputElement
        >
      ).type || ""
    );
  };

  const input = useRef<HTMLInputElement>(null);

  const copyToClipboard = async (
    e: MouseEvent<HTMLButtonElement, globalThis.MouseEvent>
  ) => {
    e.preventDefault();
    try {
      if (!navigator.clipboard) {
        input.current?.select();
        input.current?.setSelectionRange(0, 99999);
        const succ = document.execCommand("copy");
        if (succ) {
          Swal.fire({
            timer: 2000,
            text: "Copied!",
            toast: true,

            icon: "success",
            width: "15rem",
            position: "top-right",
            showConfirmButton: false,
            timerProgressBar: true,
          });
          return;
        }
        Swal.fire({
          timer: 2000,
          text: "Error, link not coppied!",
          toast: true,
          iconColor: "#cb1414",
          icon: "error",
          width: "15rem",
          position: "top-right",
          showConfirmButton: false,
          timerProgressBar: true,
        });
        return;
      }
      const link = input.current?.value;
      await navigator.clipboard.writeText(link || "");

      Swal.fire({
        timer: 2000,
        text: "Copied!",
        toast: true,

        icon: "success",
        width: "15rem",
        position: "top-right",
        showConfirmButton: false,
      });
      return;
    } catch {
      Swal.fire({
        timer: 2000,
        text: "Error, link not coppied!",
        toast: true,
        iconColor: "#cb1414",
        icon: "error",
        width: "15rem",
        position: "top-right",
        showConfirmButton: false,
        timerProgressBar: true,
      });
      return;
    }
  };

  useEffect(() => {
    if (
      props.maxLength &&
      props.maxLength > 1 &&
      props.value &&
      props.value.toString().length >= props.maxLength - 1
    ) {
      setCountClass("input__count text--error");
      return;
    }
    setCountClass("input__count");
  }, [props.value, props.maxLength]);

  return (
    <div
      className={`input__group ${containerClassName || ""} ${
        inline ? "input__group--inline" : ""
      } ${full ? "input__group--full" : ""} ${
        search ? "input__group--search" : ""
      }`}
    >
      <div
        className={`input__container ${
          !valid ? "input__container--invalid" : ""
        } ${props.disabled ? "input__container--disabled" : ""} ${
          copyToClip ? "input__container--copy" : ""
        } ${search ? "input__container--search" : ""}`}
      >
        {svgId && (
          <SvgWrapper
            id={svgId}
            className={`svg input__svg ${!valid ? "text--error" : ""} ${
              props.disabled ? "input__svg--disabled" : ""
            }`}
          />
        )}
        <div className="input__wrapper">
          {textarea ? (
            <textarea
              {...(props as DetailedHTMLProps<
                TextareaHTMLAttributes<HTMLTextAreaElement>,
                HTMLTextAreaElement
              >)}
              autoComplete="nope"
              className="input input--textarea"
              rows={4}
            ></textarea>
          ) : (
            <input
              {...(props as DetailedHTMLProps<
                InputHTMLAttributes<HTMLInputElement>,
                HTMLInputElement
              >)}
              ref={input}
              type={type}
              autoComplete="nope"
              className="input"
            />
          )}
          {!search && (
            <label
              htmlFor={props.name}
              style={{color: helpColor}}
              className={`input__label ${!valid ? "text--error" : ""} ${
                props.disabled ? "input__label--disabled" : ""
              }`}
            >
              {props.placeholder}
            </label>
          )}
        </div>
        {(
          props as DetailedHTMLProps<
            InputHTMLAttributes<HTMLInputElement>,
            HTMLInputElement
          >
        ).type === "password" ? (
          <button
            type="button"
            className="btn btn--icon input__btn"
            onMouseDown={handleShow}
            onMouseUp={handleUnshow}
            onMouseLeave={handleUnshow}
          >
            <SvgWrapper
              id={show ? svgs.eye : svgs.eyeClosed}
              className="svg input__eye"
            />
          </button>
        ) : copyToClip ? (
          <button
            type="button"
            className="btn btn--icon input__btn"
            onClick={copyToClipboard}
          >
            <SvgWrapper id={svgs.copyToClip} className="svg input__copy" />
          </button>
        ) : (
          <></>
        )}
        {search && (
          <button
            type="submit"
            className="btn btn--icon input__btn input__btn--search"
            onClick={submitFn}
          >
            <SvgWrapper id={svgs.search} className="svg input__search" />
          </button>
        )}
        {!valid ? (
          <SvgWrapper id={svgs.danger} className="svg input__svg text--error" />
        ) : (
          <></>
        )}
      </div>
      {helpText ||
      (!valid && error) ||
      props.maxLength ||
      passwordValidation ? (
        <>
          <div className="input__text">
            {helpText && valid ? (
              <span className="input__help">{helpText}</span>
            ) : error && !valid ? (
              <span className="text--error">{error}</span>
            ) : (
              <></>
            )}
            {props.maxLength && (
              <span className={countClass}>{`${
                (props.value as string)?.length || 0
              }/${props.maxLength}`}</span>
            )}
          </div>
          {passwordValidation && (
            <PasswordValidation password={(props.value as string) || ""} />
          )}
        </>
      ) : (
        <></>
      )}
    </div>
  );
};

export default Input;
