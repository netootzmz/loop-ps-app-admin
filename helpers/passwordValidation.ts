import validator from "validator";

const passwordValidation = (pass: string) => {
  const minlength = validator.isLength(pass, { min: 8, max: 20 });
  const digit = /[0-9]/g.test(pass);
  const lowercase = /[a-z]/g.test(pass);
  const uppercase = /[A-Z]/g.test(pass);
  const special = /[!#$%&()*+,\-./:;<=>?@\]\[_`{|}\\\\~]/g.test(pass);
  const nowhitespace = !/\s/g.test(pass);
  const norepeat =
    !/([a-zA-Z0-9!#$%&()*+,\-./:;<=>?@\]\[_`{|}\\\\~])\1{1,}/.test(pass);
  const norepeatnumbers = !consecutiveDigits(pass);
  const norepetcharacters = !consecutiveChars(pass);
  const noinverserepeatnumbers = !inverseConsecutiveDigits(pass);
  const noinverserepeatcharacters = !inverseConsecutiveChars(pass);

  return (
    minlength &&
    digit &&
    lowercase &&
    uppercase &&
    special &&
    nowhitespace &&
    norepeat &&
    norepeatnumbers &&
    norepetcharacters &&
    noinverserepeatnumbers &&
    noinverserepeatcharacters
  );
};

export const getPasswordValidationMessage = (
  pass: string,
  messages: {
    minLength: string;
    digit: string;
    lowercase: string;
    uppercase: string;
    special: string;
    nowhitespace: string;
    norepeat: string;
    norepeatnumbers: string;
    norepeatcharacters: string;
    noinverserepeatnumbers: string;
    noinverserepeatcharacters: string;
  }
) => {
  if (!validator.isLength(pass, { min: 8, max: 20 })) return messages.minLength;
  if (!/[0-9]/g.test(pass)) return messages.digit;
  if (!/[a-z]/g.test(pass)) return messages.lowercase;
  if (!/[A-Z]/g.test(pass)) return messages.uppercase;
  if (!/[!#$%&()*+,\-./:;<=>?@\]\[_`{|}\\\\~]/g.test(pass))
    return messages.special;
  if (/\s/g.test(pass)) return messages.nowhitespace;
  if (/([a-zA-Z0-9!#$%&()*+,\-./:;<=>?@\]\[_`{|}\\\\~])\1{1,}/.test(pass))
    return messages.norepeat;
  if (consecutiveDigits(pass)) return messages.norepeatnumbers;
  if (consecutiveChars(pass)) return messages.norepeatcharacters;
  if (inverseConsecutiveDigits(pass)) return messages.noinverserepeatnumbers;
  if (inverseConsecutiveChars(pass)) return messages.noinverserepeatcharacters;
  return "";
};

export const consecutiveDigits = (str: string) => {
  let curr: any = 0;
  let prev = 0;
  let count = 0;
  for (let i = 0; i < str.length; ++i) {
    curr = parseInt(str.split("")[i]);
    if (!isNaN(parseFloat(curr)) && isFinite(curr)) {
      if (count === 0) {
        ++count;
      } else if (prev + 1 === curr) {
        ++count;
        if (count === 2) {
          return true;
        }
      }
    }
    prev = curr;
  }
  return false;
};

export const inverseConsecutiveDigits = (str: string) => {
  let curr: any = 0;
  let prev = 0;
  let count = str.length;
  for (let i = str.length - 1; i >= 0; --i) {
    curr = parseInt(str.split("")[i]);
    if (!isNaN(parseFloat(curr)) && isFinite(curr)) {
      if (count === str.length) {
        --count;
      } else if (prev + 1 === curr) {
        --count;
        if (count !== 2) {
          return true;
        }
      }
    }
    prev = curr;
  }
  return false;
};

export const consecutiveChars = (str: string) => {
  let curr: any = 0;
  let prev = 0;
  let count = 0;
  for (let i = 0; i < str.length; ++i) {
    curr = str.charCodeAt(i);
    if (
      !isNaN(parseFloat(curr)) &&
      isFinite(curr) &&
      ((curr >= 65 && curr <= 90) ||
        (curr >= 97 && curr <= 122) ||
        (curr >= 48 && curr <= 57))
    ) {
      if (count === 0) {
        ++count;
      } else if (prev + 1 === curr) {
        ++count;
        if (count === 2) {
          return true;
        }
      }
    }
    prev = curr;
  }
  return false;
};

export const inverseConsecutiveChars = (str: string) => {
  let curr: any = 0;
  let prev = 0;
  let count = str.length;
  for (let i = str.length - 1; i >= 0; --i) {
    curr = str.charCodeAt(i);
    if (
      !isNaN(parseFloat(curr)) &&
      isFinite(curr) &&
      ((curr >= 65 && curr <= 90) ||
        (curr >= 97 && curr <= 122) ||
        (curr >= 48 && curr <= 57))
    ) {
      if (count === str.length) {
        --count;
      } else if (prev + 1 === curr) {
        --count;
        if (count !== 2) {
          return true;
        }
      }
    }
    prev = curr;
  }
  return false;
};

export default passwordValidation;
