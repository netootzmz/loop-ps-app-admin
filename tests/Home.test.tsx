import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import Input from "../components/forms/Input";
import Home from "./../pages";
import passwordValidation from "../helpers/passwordValidation";

describe("Test in <Input/>", () => {
  test("Should render Input", () => {
    const pass = passwordValidation("$mArt_158192");

    expect(pass).toBe(true);
  });
});
