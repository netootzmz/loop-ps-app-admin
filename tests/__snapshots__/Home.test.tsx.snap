// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Input/> Should render Input 1`] = `
<div
  className="input__group   "
>
  <div
    className="input__container  input__container--disabled "
  >
    <div
      className="input__wrapper"
    >
      <input
        autoComplete="nope"
        className="input"
        disabled={true}
        type=""
      />
      <label
        className="input__label  input__label--disabled"
      />
    </div>
  </div>
</div>
`;
